internal/pkg/agent/application/actions/handlers/handler_helpers.go

```diff
@@ -0,0 +1,104 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package handlers
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/component"
+)
+
+type actionCoordinator interface {
+	State() coordinator.State
+	PerformAction(ctx context.Context, comp component.Component, unit component.Unit, name string, params map[string]interface{}) (map[string]interface{}, error)
+}
+
+type upgradeCoordinator interface {
+	actionCoordinator
+	Upgrade(ctx context.Context, version string, sourceURI string, action *fleetapi.ActionUpgrade, skipVerifyOverride bool, pgpBytes ...string) error
+}
+
+type performActionFunc func(context.Context, component.Component, component.Unit, string, map[string]interface{}) (map[string]interface{}, error)
+
+type dispatchableAction interface {
+	MarshalMap() (map[string]interface{}, error)
+	Type() string
+}
+
+func dispatchActionInParallel(ctx context.Context, log *logp.Logger, action dispatchableAction, comps []component.Component, units []component.Unit, performAction performActionFunc) error {
+	if action == nil {
+		return nil
+	}
+
+	// Deserialize the action into map[string]interface{} for dispatching over to the apps
+	params, err := action.MarshalMap()
+	if err != nil {
+		return err
+	}
+
+	actionType := action.Type()
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Iterate through found components and forward the UNENROLL action
+	for i := 0; i < len(comps); i++ {
+		g.Go(func(idx int) func() error {
+			comp := comps[i]
```

Comment:
I can see by tracing through to `findMatchingUnitsByActionType`, the source of these arrays, that this is probably safe, but it's not obvious from the function definition or comments. Most arrays of components and units have no reason to be the same size, and the requirement for these parameters isn't documented.

Suggestion: define a helper like:
```go
type unitWithComponent struct {
  unit component.Unit
  component component.Component
}
```

then have `findMatchingUnits...` return an array of these, and `dispatchAction...` take that array as input, so you can just iterate directly over the list without relying on extra metadata about the list indexes. Then it's immediately obvious that the bounds checking is correct and that these values are intentionally paired.

Reply:
Done.